### PR TITLE
RSA-PSS support

### PR DESCRIPTION
**Pull request by snhenson**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change

This PR adds support for RSA-PSS public keys through a new public key method. RSA-PSS is a restricted form of RSA which only supports signing and verification using PSS padding mode. Additionally some restrictions can be placed on the PSS parameters which may be used.

WIP as the documentation may need some additional work.

### CODE REVIEW

File: crypto/evp/pmeth_lib.c [hunk 1]
```diff
@@ -364,6 +362,16 @@ int EVP_PKEY_CTX_hex2ctrl(EVP_PKEY_CTX *ctx, int cmd, const char *hex)
     OPENSSL_free(bin);
     return rv;
 }
+/* Pass a message digest to a ctrl */
```
Comment (FdaSilvaYY) [inline crypto/evp/pmeth_lib.c]: Style: Missing blank line to separate the two methods.

File: crypto/rsa/rsa_pmeth.c [hunk 2]
```diff
@@ -605,6 +646,22 @@ static int pkey_rsa_ctrl_str(EVP_PKEY_CTX *ctx,
     return -2;
 }
 
+/* Set PSS parameters when generating a key, if necessary */
+static int rsa_set_pss_param(RSA *rsa, EVP_PKEY_CTX *ctx)
+{
+    RSA_PKEY_CTX *rctx = ctx->data;
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_pmeth.c]: Style: blank line after decl.

File: crypto/rsa/rsa_pmeth.c [hunk 3]
```diff
@@ -671,3 +732,64 @@ const EVP_PKEY_METHOD rsa_pkey_meth = {
...
+ */
+
+static int pkey_pss_init(EVP_PKEY_CTX *ctx)
+{
+    RSA *rsa;
+    RSA_PKEY_CTX *rctx = ctx->data;
+    const EVP_MD *md;
+    const EVP_MD *mgf1md;
+    int min_saltlen;
+    /* Should never happen */
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_pmeth.c]: Style: blank line after declarations

File: crypto/rsa/rsa_ameth.c [hunk 4]
```diff
@@ -24,15 +24,69 @@ static int rsa_cms_decrypt(CMS_RecipientInfo *ri);
...
 #endif
 
+static RSA_PSS_PARAMS *rsa_pss_decode(const X509_ALGOR *alg);
+
+/* Set any parameters associated with pkey */
+static int rsa_param_encode(const EVP_PKEY *pkey,
+                            ASN1_STRING **pstr, int *pstrtype)
+{
+    const RSA *rsa = pkey->pkey.rsa;
+    *pstr = NULL;
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 5]
```diff
@@ -24,15 +24,69 @@ static int rsa_cms_decrypt(CMS_RecipientInfo *ri);
...
+    return 1;
+}
+
 static int rsa_pub_encode(X509_PUBKEY *pk, const EVP_PKEY *pkey)
 {
     unsigned char *penc = NULL;
     int penclen;
+    ASN1_STRING *str;
+    int strtype;
+    if (!rsa_param_encode(pkey, &str, &strtype))
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 6]
```diff
@@ -85,15 +144,19 @@ static int rsa_priv_encode(PKCS8_PRIV_KEY_INFO *p8, const EVP_PKEY *pkey)
 {
     unsigned char *rk = NULL;
     int rklen;
+    ASN1_STRING *str;
+    int strtype;
+    if (!rsa_param_encode(pkey, &str, &strtype))
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 7]
```diff
@@ -130,104 +206,39 @@ static void int_rsa_free(EVP_PKEY *pkey)
...
-
-static int rsa_pss_param_print(BIO *bp, RSA_PSS_PARAMS *pss,
-                               X509_ALGOR *maskHash, int indent)
+static int rsa_pss_param_print(BIO *bp, int pss_key, RSA_PSS_PARAMS *pss,
+                               int indent)
 {
     int rv = 0;
-    if (!pss) {
-        if (BIO_puts(bp, " (INVALID PSS PARAMETERS)\n") <= 0)
+    X509_ALGOR *maskHash = NULL;
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 8]
```diff
@@ -286,21 +298,105 @@ static int rsa_pss_param_print(BIO *bp, RSA_PSS_PARAMS *pss,
...
                          const ASN1_STRING *sig, int indent, ASN1_PCTX *pctx)
 {
-    if (OBJ_obj2nid(sigalg->algorithm) == NID_rsassaPss) {
+    if (OBJ_obj2nid(sigalg->algorithm) == EVP_PKEY_RSA_PSS) {
         int rv;
         RSA_PSS_PARAMS *pss;
-        X509_ALGOR *maskHash;
-        pss = rsa_pss_decode(sigalg, &maskHash);
-        rv = rsa_pss_param_print(bp, pss, maskHash, indent);
+        pss = rsa_pss_decode(sigalg);
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 9]
```diff
@@ -410,55 +512,37 @@ static const EVP_MD *rsa_algor_to_md(X509_ALGOR *alg)
...
 
-static ASN1_STRING *rsa_ctx_to_pss(EVP_PKEY_CTX *pkctx)
+static RSA_PSS_PARAMS *rsa_ctx_to_pss(EVP_PKEY_CTX *pkctx)
 {
     const EVP_MD *sigmd, *mgf1md;
-    RSA_PSS_PARAMS *pss = NULL;
-    ASN1_STRING *os = NULL;
     EVP_PKEY *pk = EVP_PKEY_CTX_get0_pkey(pkctx);
-    int saltlen, rv = 0;
+    int saltlen;
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 10]
```diff
@@ -410,55 +512,37 @@ static const EVP_MD *rsa_algor_to_md(X509_ALGOR *alg)
...
     }
-    pss = RSA_PSS_PARAMS_new();
+
+    return rsa_pss_params_create(sigmd, mgf1md, saltlen);
+}
+
+RSA_PSS_PARAMS *rsa_pss_params_create(const EVP_MD *sigmd,
+                                      const EVP_MD *mgf1md, int saltlen)
+{
+    RSA_PSS_PARAMS *pss = RSA_PSS_PARAMS_new();
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 11]
```diff
@@ -470,20 +554,31 @@ static ASN1_STRING *rsa_ctx_to_pss(EVP_PKEY_CTX *pkctx)
...
-        return os;
-    ASN1_STRING_free(os);
     return NULL;
 }
 
+static ASN1_STRING *rsa_ctx_to_pss_string(EVP_PKEY_CTX *pkctx)
+{
+    RSA_PSS_PARAMS *pss = rsa_ctx_to_pss(pkctx);
+    ASN1_STRING *os = NULL;
+    if (pss == NULL)
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_ameth.c [hunk 12]
```diff
@@ -580,10 +675,15 @@ static int rsa_cms_verify(CMS_SignerInfo *si)
     EVP_PKEY_CTX *pkctx = CMS_SignerInfo_get0_pkey_ctx(si);
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Style: blank line after declaration.

File: crypto/rsa/rsa_asn1.c [hunk 13]
```diff
@@ -49,20 +49,42 @@ ASN1_SEQUENCE_cb(RSAPublicKey, rsa_cb) = {
...
 } ASN1_SEQUENCE_END_cb(RSA, RSAPublicKey)
 
-ASN1_SEQUENCE(RSA_PSS_PARAMS) = {
+/* Free up maskHash */
+static int rsa_pss_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
+                      void *exarg)
+{
+    if (operation == ASN1_OP_FREE_PRE) {
+        RSA_PSS_PARAMS *pss = (RSA_PSS_PARAMS *)*pval;
+        X509_ALGOR_free(pss->maskHash);
```
Comment (snhenson) [inline crypto/rsa/rsa_asn1.c]: The callback with ASN1_OP_FREE_PRE is called prior to freeing the structure. It is used to free up any additional fields which have been added so we can rely on this behaviour.

File: crypto/asn1/standard_methods.h [hunk 14]
```diff
@@ -30,6 +30,9 @@ static const EVP_PKEY_ASN1_METHOD *standard_methods[] = {
 #ifndef OPENSSL_NO_CMAC
     &cmac_asn1_meth,
 #endif
+#ifndef OPENSSL_NO_RSA
+    &rsa_pss_asn1_meth,
+#endif
```
Comment (snhenson) [inline crypto/asn1/standard_methods.h]: It means they should be kept in ascending order of the type (i.e. NID) each method represents which corresponds to the pkey_id field of EVP_PKEY_ASN1_METHOD. This is needed because we use OBJ_bsearch() to lookup methods by pkey_id. The comment could make that clearer.

BTW if the table  gets out of order then there is a test (pkey_meth_test) which will pick it up.

File: crypto/rsa/rsa_ameth.c [hunk 15]
```diff
@@ -568,10 +632,41 @@ static int rsa_pss_to_ctx(EVP_MD_CTX *ctx, EVP_PKEY_CTX *pkctx,
...
+    if (*pmd == NULL)
+        return 0;
+    *pmgf1md = rsa_algor_to_md(pss->maskHash);
+    if (*pmgf1md == NULL)
+        return 0;
+    if (pss->saltLength) {
+        *psaltlen = ASN1_INTEGER_get(pss->saltLength);
+        if (*psaltlen < 0) {
+            RSAerr(RSA_F_RSA_PSS_GET_PARAM, RSA_R_INVALID_SALT_LENGTH);
+            return 0;
```
Comment (snhenson) [inline crypto/rsa/rsa_ameth.c]: The function rsa_algor_to_md() returns it's own error, the other cases do not.

File: crypto/rsa/rsa_ameth.c [hunk 16]
```diff
@@ -286,25 +302,109 @@ static int rsa_pss_param_print(BIO *bp, RSA_PSS_PARAMS *pss,
...
+    if (priv && x->d) {
+        if (BIO_printf(bp, "Private-Key: (%d bit)\n", mod_len) <= 0)
+            goto err;
+        str = "modulus:";
+        s = "publicExponent:";
+    } else {
+        if (BIO_printf(bp, "Public-Key: (%d bit)\n", mod_len) <= 0)
+            goto err;
+        str = "Modulus:";
+        s = "Exponent:";
```
Comment (mattcaswell) [inline crypto/rsa/rsa_ameth.c]: Are the upper case label names vs lower case label names significant?

File: crypto/rsa/rsa_ameth.c [hunk 17]
```diff
@@ -470,20 +563,32 @@ static ASN1_STRING *rsa_ctx_to_pss(EVP_PKEY_CTX *pkctx)
...
+static ASN1_STRING *rsa_ctx_to_pss_string(EVP_PKEY_CTX *pkctx)
+{
+    RSA_PSS_PARAMS *pss = rsa_ctx_to_pss(pkctx);
+    ASN1_STRING *os = NULL;
+
+    if (pss == NULL)
+        return NULL;
+
+    if (!ASN1_item_pack(pss, ASN1_ITEM_rptr(RSA_PSS_PARAMS), &os)) {
+        ASN1_STRING_free(os);
```
Comment (mattcaswell) [inline crypto/rsa/rsa_ameth.c]: And here...not sure why this free is necessary.

File: crypto/rsa/rsa_pmeth.c [hunk 18]
```diff
@@ -35,22 +35,30 @@ typedef struct {
...
 } RSA_PKEY_CTX;
 
+/* True if PSS parameters are restricted */
+#define rsa_pss_param(rctx) (rctx->min_saltlen != -1)
+
 static int pkey_rsa_init(EVP_PKEY_CTX *ctx)
 {
-    RSA_PKEY_CTX *rctx;
-    rctx = OPENSSL_zalloc(sizeof(*rctx));
+    RSA_PKEY_CTX *rctx = OPENSSL_zalloc(sizeof(*rctx));
```
Comment (mattcaswell) [inline crypto/rsa/rsa_pmeth.c]: Blank line after decl

**Comment by mattcaswell** [comment]
There is a travis failure which seems to indicate one of the cms fuzz tests has failed.

File: crypto/rsa/rsa_pmeth.c [hunk 19]
```diff
@@ -35,22 +35,31 @@ typedef struct {
...
+    int min_saltlen;
     /* Temp buffer */
     unsigned char *tbuf;
     /* OAEP label */
     unsigned char *oaep_label;
     size_t oaep_labellen;
 } RSA_PKEY_CTX;
 
+/* True if PSS parameters are restricted */
+#define rsa_pss_param(rctx) (rctx->min_saltlen != -1)
```
Comment (snhenson) [inline crypto/rsa/rsa_pmeth.c]: Done in subsequent commit.

File: include/openssl/rsa.h [hunk 20]
```diff
@@ -86,46 +86,50 @@ extern "C" {
...
+                          EVP_PKEY_CTRL_RSA_KEYGEN_PUBEXP, 0, pubexp)
 
 # define  EVP_PKEY_CTX_set_rsa_mgf1_md(ctx, md)  \
-                EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_RSA, \
-                        EVP_PKEY_OP_TYPE_SIG | EVP_PKEY_OP_TYPE_CRYPT, \
-                                EVP_PKEY_CTRL_RSA_MGF1_MD, 0, (void *)md)
+        RSA_pkey_ctx_ctrl(ctx, EVP_PKEY_OP_TYPE_SIG | EVP_PKEY_OP_TYPE_CRYPT, \
+                          EVP_PKEY_CTRL_RSA_MGF1_MD, 0, (void *)md)
+
+# define  EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md(ctx, md)  \
```
Comment (snhenson) [inline include/openssl/rsa.h]: Yes that's one reason why this is WIP. Documentation added to subsequent commit.

File: crypto/rsa/rsa_ameth.c (hunk 16 above)
Comment (snhenson) [inline crypto/rsa/rsa_ameth.c]: No, that's just to be consistent with the previous format.

File: crypto/rsa/rsa_ameth.c [hunk 21]
```diff
@@ -24,15 +24,71 @@ static int rsa_cms_decrypt(CMS_RecipientInfo *ri);
...
+    /* If no PSS parameters we omit parameters entirely */
+    if (rsa->pss == NULL) {
+        *pstrtype = V_ASN1_UNDEF;
+        return 1;
+    }
+    /* Encode PSS parameters */
+    if (!ASN1_item_pack(rsa->pss, ASN1_ITEM_rptr(RSA_PSS_PARAMS), pstr)) {
+        ASN1_STRING_free(*pstr);
+        *pstr = NULL;
+        return 0;
```
Comment (snhenson) [inline crypto/rsa/rsa_ameth.c]: No it isn't needed and we should also check != NULL as it returns a pointer.

File: doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+It has associated private key and public key formats.
+
+This algorithm shares several control operations with the B<RSA> algorithm
+but with some restrictions described below.
+
+=head1 SIGNING AND VERIFICATION
+
+Siging and verification is similar to the B<RSA> algorithm except the
+padding mode is always PSS. If the key used restricts parameters then
```
Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: "If the key used restricts parameters" is easy to start parsing incorrectly in a way that "If the key in use has parameter restrictions" is not.

File: doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+the corresponding signature parameters are set to the restrictions:
+for example if the key can only be used with digest SHA256, MGF1 SHA256
```
Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: Maybe put a comma after "for example"?

File: doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+and minimum salt length 32 then the digest, MGF1 digest and salt length
+will be set to SHA256, SHA256 and 32 respectively.
+
+The macro EVP_PKEY_CTX_set_rsa_padding() is supported but an error is
+returned if an attempt is made to set the padding mode to anything other
+than B<PSS>. It is otherwise similar to the B<RSA> version.
+
+The EVP_PKEY_CTX_set_rsa_pss_saltlen() macro is used to set the salt length.
+If the key is restricted then an error is returned if an attempt is made
```
Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: "is restricted" makes me think of a restricted area in a library or something; how about "has usage restrictions"?  (Shows up again later in the file, at least once across a line boundary.)

File: doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+to set the salt below the minimum value. It is otherwise similar to the
```
Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: set the salt, set the salt length, or set a salt whose length is?

File: doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+similar to the B<RSA> versions.
+
+=head1 KEY GENERATION
+
+As with RSA key generation the EVP_PKEY_CTX_set_rsa_rsa_keygen_bits()
+and EVP_PKEY_CTX_set_rsa_keygen_pubexp() macros are supported for RSA PSS:
+they have exactly the same meaning as for the RSA algorithm.
+
+Optional parameter restrictions can be specified when generating a PSS key
+by default no parameter restrictions are placed on a generate key. If any
```
Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: This feels like it should be a new sentence.

Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: s/generate/generated/

File: doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+restrictions are set (using the macros described below) then B<all>
+parameters are restricted: e.g. setting a minimum salt length also restricts
```
Comment (kaduk) [inline doc/man3/EVP_PKEY_CTX_set_rsa_pss_keygen_mgf1_md.pod]: Comma after e.g.

**Comment by snhenson** [comment]
I think that addressed all the review comments so far and it passes travis now though the code wasn't changed significanltly. Took out of WIP.

File: crypto/rsa/rsa_ameth.c [hunk 28]
```diff
@@ -286,25 +301,109 @@ static int rsa_pss_param_print(BIO *bp, RSA_PSS_PARAMS *pss,
...
                          const ASN1_STRING *sig, int indent, ASN1_PCTX *pctx)
 {
-    if (OBJ_obj2nid(sigalg->algorithm) == NID_rsassaPss) {
+    if (OBJ_obj2nid(sigalg->algorithm) == EVP_PKEY_RSA_PSS) {
         int rv;
-        RSA_PSS_PARAMS *pss;
-        X509_ALGOR *maskHash;
-        pss = rsa_pss_decode(sigalg, &maskHash);
-        rv = rsa_pss_param_print(bp, pss, maskHash, indent);
+        RSA_PSS_PARAMS *pss = rsa_pss_decode(sigalg);
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_ameth.c]: Blank line after declarations

File: crypto/rsa/rsa_pmeth.c [hunk 29]
```diff
@@ -572,23 +610,32 @@ static int pkey_rsa_ctrl_str(EVP_PKEY_CTX *ctx,
...
+        if (strcmp(type, "rsa_pss_keygen_mgf1_md") == 0)
+            return EVP_PKEY_CTX_md(ctx, EVP_PKEY_OP_KEYGEN,
+                                   EVP_PKEY_CTRL_RSA_MGF1_MD, value);
+
+        if (strcmp(type, "rsa_pss_keygen_md") == 0)
+            return EVP_PKEY_CTX_md(ctx, EVP_PKEY_OP_KEYGEN,
+                                   EVP_PKEY_CTRL_MD, value);
+
+        if (strcmp(type, "rsa_pss_keygen_saltlen") == 0) {
+            int saltlen;
```
Comment (FdaSilvaYY) [inline crypto/rsa/rsa_pmeth.c]: Blank line after declaration.
Could be merged with atoi call ?

File: crypto/evp/pmeth_lib.c [hunk 30]
```diff
@@ -365,6 +363,17 @@ int EVP_PKEY_CTX_hex2ctrl(EVP_PKEY_CTX *ctx, int cmd, const char *hex)
     return rv;
 }
 
+/* Pass a message digest to a ctrl */
+int EVP_PKEY_CTX_md(EVP_PKEY_CTX *ctx, int optype, int cmd, const char *md)
+{
+    const EVP_MD *m;
```
Comment (FdaSilvaYY) [inline crypto/evp/pmeth_lib.c]: Blank line after declaration

**Comment by snhenson** [comment]
Ping @mattcaswell are you OK with this now?

File: apps/opt.c [hunk 31]
```diff
@@ -358,7 +358,7 @@ int opt_imax(const char *value, intmax_t *result)
 {
     int oerrno = errno;
     intmax_t m;
-    char *endp;
```
Comment (FdaSilvaYY) [inline apps/opt.c]: Why this change ?

Comment (snhenson) [inline apps/opt.c]: Oops, removed.

**Comment by snhenson** [comment]
Pushed now, thanks everyone!
